File: code/141.py
'''
Author: lmio 2091319361@qq.com
Date: 2023-07-07 17:57:35
LastEditors: lmio 2091319361@qq.com
Description: 141. 环形链表
'''

from typing import Optional

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

def hasCycle(head: Optional[ListNode]) -> bool:
    if not head or not head.next:
        return False
    slow, fast = head, head.next
    while slow and fast:
        if slow == fast:
            return True
        slow = slow.next
        fast = fast.next
        if fast:
            fast = fast.next
    return False

File: code_cpp/141.c
/*
 * @Author: lmio 2091319361@qq.com
 * @Date: 2024-05-06 14:30:08
 * @LastEditors: lmio 2091319361@qq.com
 * @Description: 141.环形链表
 */
#include<stdbool.h>
struct ListNode {
    int val;
    struct ListNode *next;
};

bool hasCycle(struct ListNode *head) {
    if (!head) {
        return false;
    }
    // 头到环长度为a，环长度为b slow走了x，则fast走了2x
    // 某时刻相遇，则a+nb == 2*(a+mb)
    struct ListNode* fast = head->next;
    struct ListNode* slow = head;
    while (fast && slow) {
        if (fast == slow) {
            return true;
        }
        slow = slow->next;
        fast = fast->next;
        if (fast) {
            fast = fast->next;
        }
    }
    return false;
}